File: mobile/ios/OTravelz/Features/Notifications/TransitReminderCalculator.swift
import Foundation

/// Pure deterministic calculator for transit reminder trigger timestamps on iOS.
/// Enforces Indian Standard Time (Asia/Kolkata, UTC+05:30) timezone arithmetic.
public enum TransitReminderCalculator {

    public static let istTimeZone: TimeZone = TimeZone(identifier: "Asia/Kolkata") ?? TimeZone(secondsFromGMT: 19800)!

    /// Calculates the trigger epoch millisecond timestamp for a reminder.
    /// Returns nil if the departure has already passed or if time string is malformed.
    public static func calculateTriggerEpochMs(
        departureTime: String,
        offsetMinutes: Int,
        referenceDate: Date = Date(),
        timeZone: TimeZone = istTimeZone
    ) -> Int64? {
        guard offsetMinutes >= 0 else { return nil }

        let parts = departureTime.trimmingCharacters(in: .whitespaces).split(separator: ":")
        guard parts.count == 2,
              let hour = Int(parts[0]),
              let minute = Int(parts[1]),
              hour >= 0 && hour <= 23,
              minute >= 0 && minute <= 59 else {
            return nil
        }

        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = timeZone

        var components = calendar.dateComponents([.year, .month, .day], from: referenceDate)
        components.hour = hour
        components.minute = minute
        components.second = 0

        guard let departureDate = calendar.date(from: components) else {
            return nil
        }

        let departureEpochMs = Int64(departureDate.timeIntervalSince1970 * 1000)
        let triggerEpochMs = departureEpochMs - Int64(offsetMinutes * 60 * 1000)
        let nowEpochMs = Int64(referenceDate.timeIntervalSince1970 * 1000)

        // Past departure rejection
        if triggerEpochMs <= nowEpochMs {
            return nil
        }

        return triggerEpochMs
    }

    /// Checks if a departure has already passed relative to reference date.
    public static func isDeparturePassed(
        departureTime: String,
        referenceDate: Date = Date(),
        timeZone: TimeZone = istTimeZone
    ) -> Bool {
        return calculateTriggerEpochMs(
            departureTime: departureTime,
            offsetMinutes: 0,
            referenceDate: referenceDate,
            timeZone: timeZone
        ) == nil
    }
}

File: mobile/android/src/main/kotlin/com/otravelz/android/notifications/TransitReminderCalculator.kt
package com.otravelz.android.notifications

import java.time.LocalDate
import java.time.LocalTime
import java.time.ZoneId
import java.time.ZonedDateTime
import java.time.format.DateTimeParseException

/**
 * Pure deterministic calculator for transit reminder trigger timestamps.
 * Enforces IST (Asia/Kolkata, UTC+05:30) timezone arithmetic and past-departure rejection.
 */
object TransitReminderCalculator {

    val ZONE_IST: ZoneId = ZoneId.of("Asia/Kolkata")

    /**
     * Calculates the exact epoch millisecond trigger timestamp for a reminder.
     *
     * @param departureTime Time string formatted as "HH:mm" in IST.
     * @param offsetMinutes Minutes before departure to trigger reminder (e.g. 10, 15, 30).
     * @param targetDate The date of the departure (defaults to today in IST).
     * @param nowEpochMs The current reference epoch timestamp (defaults to current system time).
     * @param zoneId Timezone of the timetable (strictly IST).
     * @return Trigger epoch milliseconds, or null if the departure is malformed or already passed.
     */
    fun calculateTriggerEpochMs(
        departureTime: String,
        offsetMinutes: Int,
        targetDate: LocalDate = LocalDate.now(ZONE_IST),
        nowEpochMs: Long = System.currentTimeMillis(),
        zoneId: ZoneId = ZONE_IST
    ): Long? {
        if (offsetMinutes < 0) return null

        val localTime = try {
            val parts = departureTime.trim().split(":")
            if (parts.size != 2) return null
            val hour = parts[0].toIntOrNull() ?: return null
            val minute = parts[1].toIntOrNull() ?: return null
            if (hour !in 0..23 || minute !in 0..59) return null
            LocalTime.of(hour, minute)
        } catch (e: Exception) {
            return null
        }

        val zonedDateTime = ZonedDateTime.of(targetDate, localTime, zoneId)
        val departureEpochMs = zonedDateTime.toInstant().toEpochMilli()
        val triggerEpochMs = departureEpochMs - (offsetMinutes * 60 * 1000L)

        // Strict rejection: if trigger time is in the past, cannot schedule
        if (triggerEpochMs <= nowEpochMs) {
            return null
        }

        return triggerEpochMs
    }

    /**
     * Checks if a departure has already passed relative to the current IST time.
     */
    fun isDeparturePassed(
        departureTime: String,
        targetDate: LocalDate = LocalDate.now(ZONE_IST),
        nowEpochMs: Long = System.currentTimeMillis(),
        zoneId: ZoneId = ZONE_IST
    ): Boolean {
        val triggerMs = calculateTriggerEpochMs(
            departureTime = departureTime,
            offsetMinutes = 0,
            targetDate = targetDate,
            nowEpochMs = nowEpochMs,
            zoneId = zoneId
        )
        return triggerMs == null
    }
}
